mono/profiler/log.c

```diff
@@ -1347,8 +1349,13 @@ gc_event (MonoProfiler *profiler, MonoProfilerGCEvent ev, uint32_t generation)
 		 * a shutdown heapshot. Either way, a manually triggered heapshot
 		 * overrides any decision we made in the switch above.
 		 */
-		if (mono_atomic_load_i32 (&log_profiler.heapshot_requested))
+		if (is_major && is_serial && mono_atomic_load_i32 (&log_profiler.heapshot_requested)) {
 			log_profiler.do_heap_walk = TRUE;
+		} else if (log_profiler.do_heap_walk && (!is_major || !is_serial)) {
+			/* Do a heap walk later, when we get invoked from the finalizer in serial mode */
```

Comment:
We should probably cycle the finalizer thread here? I.e. invoke `trigger_heapshot`? I know it can't do much *immediately* since this thread is currently doing a collection, but at least the finalizer thread is guaranteed to do the serial + major GC as soon as this thread finishes the collection.